Functional tests for Iceberg on cloud AWS + Unity Catalog

rpdevmp opened the pull request:
Functional tests for Iceberg on cloud AWS + Unity Catalog

- Reads Iceberg specific params and secrets from globals.json (new params were added by DEVPROD-2888) those would be used for local setups and by Buildkite schedules

- Uses public APIs that would be used by customers

- APIs for secret management on dataplane (new feature) @alenkacz FYI

@nvartolomei
- Includes a light-weight Databricks client to work with hosted setup I see that you merged your CDT PR several hours ago.. In your PR it is CDT specific (storage and a few other things not cloud ready), but some items are overlapping as we discussed, so we could consolidate those later to work for CDT and cloud. I would like to merge this client as it was already tested many times for cloud setups

- Uses S3 bucket object storage (AWS provider)
- Sets everything up and produces data to topics
 
DRAFT since
Need to add verification of object storage and queries and cleanups

Testing was done many times, these functions work and including error handling

## Backports Required

- [X ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes
* none

rpdevmp committed:
functional tests for Iceberg on cloud AWS + Unity Catalog

Added Redpanda secret management APIs and updated databricks client

removed databricks_client, since another pr already got merged

Adding dataplane URL fetch, redpanda secrets management

code review changes

secpanda commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devex-c3w/pr-checks/74e3dbaa-ac06-4f34-8167-6e78d810b016?refs=93947e2a1a85815854af6469bce8e86234c0ed3a&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devex-c3w/pr-checks/74e3dbaa-ac06-4f34-8167-6e78d810b016/license?refs=93947e2a1a85815854af6469bce8e86234c0ed3a&source=prChecksComment)

secpanda commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devprod/pr-checks/eaaa9e8c-60fa-4954-a449-dd323adc4342?refs=93947e2a1a85815854af6469bce8e86234c0ed3a&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devprod/pr-checks/eaaa9e8c-60fa-4954-a449-dd323adc4342/license?refs=93947e2a1a85815854af6469bce8e86234c0ed3a&source=prChecksComment)

rpdevmp commented:
I will convert it to open PR once I add query checks, object storage verification and cleanups

secpanda commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/cloud-1t5/pr-checks/7cfc0b60-7145-499b-9c19-b46e1b83f13d?refs=93947e2a1a85815854af6469bce8e86234c0ed3a&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/cloud-1t5/pr-checks/7cfc0b60-7145-499b-9c19-b46e1b83f13d/license?refs=93947e2a1a85815854af6469bce8e86234c0ed3a&source=prChecksComment)

rpdevmp commented:
@nvartolomei, this PR uses databricks workspace that we already merged, added cloud specific Redpanda secret management APIs and cluster dataplane fetching. This has been tested multiple times and all secret management items works. Console API to create a secret, dataplane API got pass that secret into Redpanda cloud config and then it is able to authenticate with databricks workspace..

Let's make sure we merge this PR, so I could start adding other code to these tests and have schedules running. This PR is ready for merge (once CI checks pass)

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/65674-->
<!--a84f5f38-b64e-488a-8494-50ff6be0b7cc-->
<!--title-->
## Retry command for [Build#65674](https://buildkite.com/redpanda/redpanda/builds/65674)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/scaling_up_test.py::ScalingUpTest.test_scaling_up_with_recovered_topic
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--9fba5d8a-23e6-42cc-9e60-176a61128654-->
<!--TEST RESULTS-->
<!--f052647f-7e09-4e2c-a6bb-e35ad2f13b1f-->
<!--TEST RESULTS-->
<!--c6eaf523-b181-4cd9-8e6e-51b474b6e26a-->
<!--TEST RESULTS-->
<!--25d3f201-4c09-4093-855c-1064cbd57184-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65674">build#65674</a></summary>


| test_id                                                                                                                                                                                                                                           | test_kind   | job_url                                                                                   | test_status   | passed   |
|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=False.with_tiered_storage=False.with_iceberg=True.with_chunked_compaction=True.cloud_storage_type=CloudStorageType.S3 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65674#0196adc2-ae2e-4c2c-911d-c52b49a4edb9 | FLAKY         | 11/12    |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=False.with_iceberg=False.with_chunked_compaction=True.cloud_storage_type=CloudStorageType.S3 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65674#0196adc2-ae2e-4c2c-911d-c52b49a4edb9 | FLAKY         | 10/11    |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic                                                                                                                                                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65674#0196adc2-ae2c-4d50-9335-84fe1005a13f | FLAKY         | 10/21    |
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False                                                                                                                                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65674#0196adc2-ae2e-4c2c-911d-c52b49a4edb9 | FLAKY         | 8/10     |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65776">build#65776</a></summary>


| test_class                        | test_method                | test_arguments                                                                                                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:----------------------------------|:---------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| PartitionBalancerTest             | test_unavailable_nodes     |                                                                                                                                                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65776#0196b7f0-7b7b-4de7-8550-2ce92c0e67be | FLAKY         | 10/21    | upstream reliability is '72.31040564373897'. current run reliability is '47.61904761904761'. drift is 24.69136 and the allowed drift is set to 50. The test should PASS |
| PartitionForceReconfigurationTest | test_basic_reconfiguration | {"acks": -1, "controller_snapshots": false, "restart": false}                                                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65776#0196b7e4-5653-4816-9c80-43312b1155d6 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               |
| RandomNodeOperationsTest          | test_node_operations       | {"cloud_storage_type": 1, "enable_failures": false, "mixed_versions": false, "with_chunked_compaction": true, "with_iceberg": true, "with_tiered_storage": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65776#0196b7f0-7b79-474c-a642-83ce93787050 | FLAKY         | 19/21    | upstream reliability is '98.4375'. current run reliability is '90.47619047619048'. drift is 7.96131 and the allowed drift is set to 50. The test should PASS            |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65862">build#65862</a></summary>


| test_class                   | test_method                                       | test_arguments                                                                                                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:-----------------------------|:--------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageTimingStressTest | test_cloud_storage_with_partition_moves           | {"cleanup_policy": "compact,delete"}                                                                                                                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65862#0196c75b-e59d-430c-9333-6edc633ae158 | FLAKY         | 20/21    | upstream reliability is '96.05263157894737'. current run reliability is '95.23809523809523'. drift is 0.81454 and the allowed drift is set to 50. The test should PASS  |
| RestCatalogConnectionTest    | test_redpanda_connection_to_rest_catalog          | {"cloud_storage_type": 1}                                                                                                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65862#0196c75b-e59b-4bd9-a1f3-3a07169c9f4b | FLAKY         | 19/21    | upstream reliability is '97.26027397260275'. current run reliability is '90.47619047619048'. drift is 6.78408 and the allowed drift is set to 50. The test should PASS  |
| PartitionReassignmentsTest   | test_add_partitions_with_inprogress_reassignments |                                                                                                                                                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65862#0196c766-82c5-4d84-acb6-0f1820a221c2 | FLAKY         | 19/21    | upstream reliability is '89.92248062015504'. current run reliability is '90.47619047619048'. drift is -0.55371 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest     | test_node_operations                              | {"cloud_storage_type": 1, "enable_failures": false, "mixed_versions": false, "with_chunked_compaction": true, "with_iceberg": true, "with_tiered_storage": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65862#0196c75b-e59b-4bd9-a1f3-3a07169c9f4b | FLAKY         | 20/21    | upstream reliability is '96.89655172413794'. current run reliability is '95.23809523809523'. drift is 1.65846 and the allowed drift is set to 50. The test should PASS  |
| RandomNodeOperationsTest     | test_node_operations                              | {"cloud_storage_type": 1, "enable_failures": false, "mixed_versions": false, "with_chunked_compaction": true, "with_iceberg": false, "with_tiered_storage": true} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65862#0196c75b-e59b-4bd9-a1f3-3a07169c9f4b | FLAKY         | 20/21    | upstream reliability is '99.23076923076923'. current run reliability is '95.23809523809523'. drift is 3.99267 and the allowed drift is set to 50. The test should PASS  |
| TestTieredStoragePause       | test_safe_pause_resume                            | {"allow_gaps_cluster_level": true, "allow_gaps_topic_level": true}                                                                                                | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65862#0196c75b-e59d-430c-9333-6edc633ae158 | FLAKY         | 20/21    | upstream reliability is '98.49624060150376'. current run reliability is '95.23809523809523'. drift is 3.25815 and the allowed drift is set to 50. The test should PASS  |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65900">build#65900</a></summary>


| test_class                   | test_method                                       | test_arguments                                                                                                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:-----------------------------|:--------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| AccessControlListTestUpgrade | test_upgrade_sasl                                 |                                                                                                                                                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65900#0196c998-e2f7-4cc7-be31-33fb555ee734 | FLAKY         | 20/21    | upstream reliability is '98.46153846153847'. current run reliability is '95.23809523809523'. drift is 3.22344 and the allowed drift is set to 50. The test should PASS  |
| PartitionBalancerTest        | test_unavailable_nodes                            |                                                                                                                                                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65900#0196c9a6-1384-4d89-a9f4-c069650a9a1a | FLAKY         | 11/21    | upstream reliability is '71.05263157894737'. current run reliability is '52.38095238095239'. drift is 18.67168 and the allowed drift is set to 50. The test should PASS |
| PartitionReassignmentsTest   | test_add_partitions_with_inprogress_reassignments |                                                                                                                                                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65900#0196c9a6-1384-4d89-a9f4-c069650a9a1a | FLAKY         | 20/21    | upstream reliability is '89.03846153846153'. current run reliability is '95.23809523809523'. drift is -6.19963 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest     | test_node_operations                              | {"cloud_storage_type": 1, "enable_failures": false, "mixed_versions": false, "with_chunked_compaction": true, "with_iceberg": true, "with_tiered_storage": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65900#0196c9a6-1382-4ef9-8004-aaeca875846d | FLAKY         | 20/21    | upstream reliability is '96.65551839464884'. current run reliability is '95.23809523809523'. drift is 1.41742 and the allowed drift is set to 50. The test should PASS  |
| RandomNodeOperationsTest     | test_node_operations                              | {"cloud_storage_type": 1, "enable_failures": false, "mixed_versions": false, "with_chunked_compaction": true, "with_iceberg": false, "with_tiered_storage": true} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65900#0196c9a6-1382-4ef9-8004-aaeca875846d | FLAKY         | 20/21    | upstream reliability is '99.15966386554622'. current run reliability is '95.23809523809523'. drift is 3.92157 and the allowed drift is set to 50. The test should PASS  |
| UpgradeBackToBackTest        | test_upgrade_with_all_workloads                   | {"single_upgrade": false}                                                                                                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65900#0196c998-e2f8-40f2-b853-5c691197c1ab | FLAKY         | 11/21    | upstream reliability is '84.34782608695653'. current run reliability is '52.38095238095239'. drift is 31.96687 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

nvartolomei commented on tests/rptest/redpanda_cloud_tests/iceberg_catalogs_test.py:
```diff
@@ -0,0 +1,134 @@
+    @cluster(num_nodes=1)
+    def test_databricks_basic(self):
+        # Get globals.json map
+        globals = self._ctx.globals
```
why you are not using the typed class for accessing globals?

```
from rptest.context.databricks import DatabricksContext
dbx_ctx = DatabricksContext.from_context(self._ctx)

# on dbx_ctx now you have all you need and it is typed
# i.e. `dbx_ctx.iceberg_rest_url`,  `dbx_ctx.sql_warehouse_path`, etc
```

credentials might be a bit trickier

```
creds = Credentials.from_context(self._ctx)
assert isinstance(creds, OauthCredentials), 'databricks must be configured with oauth credentials for this test'

# now you should have typed access to `creds.client_id`
```

nvartolomei commented on tests/rptest/services/redpanda_cloud.py:
```diff
@@ -580,6 +581,9 @@ def _update_live_cluster_info(self):
         self.current.zones = _c['spec']['zones']
         self.current.network_id = _c['spec']['networkId']
         self.current.network_cidr = _c['spec']['network']['networkCidr']
+        self.current.dataplane_public_api_url = _c.get(
```
unused?

nvartolomei commented on tests/rptest/services/redpanda_cloud.py:
```diff
@@ -153,6 +153,7 @@ class LiveClusterParams:
     name: str | None = None
     last_status: str = ""
     consoleUrl: str = ""
+    dataplane_public_api_url: str = ""
```
if used this is `str | None` based on the assignment below

nvartolomei commented on tests/rptest/services/redpanda_cloud.py:
```diff
@@ -1562,3 +1567,49 @@ def set_cluster_config_overrides(self, cluster_id, config_values):
                                        json=payload)
+
+    # Update cluster properties using Public API
+    def update_cluster_property_public(self, cluster_id, property_key,
```
make this accept a map of Map[str, str]? then you'll be able to replace the _http_patch call from within the test with this. i.e. the one setting iceberg endpoint and such

nvartolomei commented:
Also, please squash the commits. The adding and removing of the previous client is not necessary in the tree and will just confuse everyone who will get blame the code.

rpdevmp commented:
Thank you for review @nvartolomei I added new changes based on the context code that was added (I was reading from globals for sometime before that was available), now updated everything to use that and removed client that I had and squashed changes.

Hopefully all checks would pass. Last time there were unrelated failures.

If looks good, let's merge this, so I could add to schedule and then could add other code. Need to get this in to verify infra setup

nvartolomei reviewed (changes requested):
requested changes privately. the test is broken

nvartolomei commented on tests/rptest/services/databricks_workspace.py:
```diff
@@ -83,12 +88,17 @@ def create_catalog(self, bucket: str) -> "DatabricksCatalogInfo":
+            self.logger.debug("SQL connection established successfully.")
+        except Exception as e:
+            self.logger.error(f"Error establishing SQL connection: {e}")
+            return  # Exit if the connection fails
```
why we don't reraise?

nvartolomei commented on tests/rptest/redpanda_cloud_tests/iceberg_catalogs_test.py:
```diff
@@ -0,0 +1,150 @@
+        # Construct the payload for the request
+        properties = {
+            "iceberg_rest_catalog_endpoint": iceberg_rest_catalog_endpoint,
+            "iceberg_rest_catalog_authentication_mode": "oauth2" if isinstance(dbx_ctx.credentials, OauthCredentials) else "pat",
```
i'm not sure if this is supported at all in cloud but pat is not a valid value, bearer is

nvartolomei commented on tests/rptest/redpanda_cloud_tests/iceberg_catalogs_test.py:
```diff
@@ -0,0 +1,150 @@
+        self.logger.debug(f"Create secret response: {create_resp}")
+
+        iceberg_rest_catalog_endpoint = dbx_ctx.iceberg_rest_url
+        iceberg_rest_catalog_warehouse = dbx_ctx.sql_warehouse_path
```
`iceberg_rest_catalog_warehouse` should be `catalog_info.name`

dbx_ctx.sql_warehouse_path relates to https://www.databricks.com/product/databricks-sql and is not something that redpanda needs to know about or knows how to use it at all.

nvartolomei commented:
The tests failed because nessie smoke test can't run. Does this branch need a rebase? 🤔 also python linting is failing. files need to be formatted properly `task lint:python`

rpdevmp committed:
minor fixes and error handling
